packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+  affirmative: PP[];
+  /** List of only the negated patterns in `patterns` */
+  negated: Extract<PP, NegatedPattern>[];
+};
+
+/**
+ * The global (top-level) config and each local (output target) config can have
+ * patterns which are separable into "watch" (always takes precedence), "documents",
+ * and "schemas". This type can hold sorted versions of these patterns.
+ */
+type PatternSet = {
+  watch: SortedPatterns;
+  documents: SortedPatterns;
+  schemas: SortedPatterns;
+};
+
+/**
+ * Return true `true` if `relativeCandidatePath` matches at least one pattern
+ * in `patterns` AND is not negated by any negation pattern in `patterns`
+ *
+ * @param relativeCandidatePath A relative path to evaluate against the patterns
+ */
+const pathMatches = (relativeCandidatePath: string, patterns: string[]) => {
+  if (isAbsolute(relativeCandidatePath)) {
+    throw new Error('pathMatches should only be called with relative path');
+  }
+
+  // We use the default micromatch export because it is the only function that
+  // implements the compound matching logic as desired (.isMatch does not remove negations).
+  // This means we don't get a nice boolean result and must check array length instead
+  const match = mm([relativeCandidatePath], patterns, { cwd: process.cwd() }).length === 1;
+
+  return match;
+};
+
+/**
+ * Filter the provided list of patterns to include only "affirmative" (non-negated) patterns.
+ *
+ * @param patterns List of micromatch patterns (or paths) to filter
+ */
+const onlyAffirmativePatterns = (patterns: string[]) => {
+  return patterns.filter(pattern => !mm.scan(pattern).negated);
+};
+
+/**
+ * Filter the provided list of patterns to include only negated patterns.
+ *
+ * @param patterns List of micromatch patterns (or paths) to filter
+ */
+const onlyNegatedPatterns = (patterns: string[]) => {
+  return patterns.filter(pattern => mm.scan(pattern).negated);
+};
+
+/**
+ * Given a list of negated patterns, invert them by removing their negation prefix
+ *
+ * If there is a non-negated pattern in the list, throw an error, because this
+ * function should only be called after filtering the list to be only negated patterns
+ *
+ * @param patterns List of negated micromatch patterns
+ */
+const invertNegatedPatterns = (patterns: string[]) => {
+  return patterns.map(pattern => {
+    const scanned = mm.scan(pattern);
+    if (!scanned.negated) {
+      throw new Error(`onlyNegatedPatterns got a non-negated pattern: ${pattern}`);
+    }
+
+    // SAFETY: Defensive programming, it's possible some picomatch/micromatch logic
+    // could call a patter negated but without its prefix !. We don't know how to
+    // handle this, but it's not worth crashing over, so warn the user and leave the pattern untransformed
+    if (scanned.negated && scanned.prefix !== '!') {
+      log(`${logSymbols.warning} Got confusing pattern, appears negated but without ! prefix: ${pattern}`);
+      return pattern;
+    }
```

Comment:
I couldn't find any instance of a case where a pattern is both `negated` and does not have a prefix of `!`, but just to be safe I added this warning here and made it fallback to the untransformed pattern, rather than risking slicing the first character of the pattern. This means the worst failure mode will be some missed rebuilds, and at least the user will see a warning.

Reply:
After writing test cases, I did find an instance of where a pattern is both `negated` and has a prefix other than `!`:

```js
mm.scan("!./foo/bar/never-watch.graphql").prefix === '!./'
```

So I removed this defensive programming check in favor of slicing off the length of the prefix when the pattern is negated.